### DISCUSSION TITLE
Android : E/VideoEncoder: Valid encoder not found : AMS RTMP

### DISCUSSION THREAD

**Discussion by hemantkathuria** [pedroSG94/RootEncoder#972] [Q&A] [answered]
Hello,
I am trying to stream android phone camera and microphone to azure media service. I am trying the ExampleRTMP activity.
I have given the rtmp URL exposed by the Azure Media Service as input in ExampleRTMP activity.
But it keeps coming to the below highlighted line. I am testing it in Google Pixel Emulator.

if (rtmpCamera1.isRecording()
              || rtmpCamera1.prepareAudio() && rtmpCamera1.prepareVideo()) {
            button.setText(R.string.stop_button);
            rtmpCamera1.startStream(etUrl.getText().toString());
          } else {
            **Toast.makeText(this, "Error preparing stream, This device cant do it",
                Toast.LENGTH_SHORT).show();**
          }
In the console log, i can see the below messages
I/AudioEncoder: prepared
**I/VideoEncoder: 0 encoders found
E/VideoEncoder: Valid encoder not found**

**Comment by pedroSG94** [comment]
This seems that you haven't h264 encoder.
Can you show me log from this method?:
```java
for (String codec: CodecUtil.showAllCodecsInfo()) {
      Log.i("Pedro", codec);
}
```

**Reply by hemantkathuria** [reply]
Btw, on the live device, i saw some success. I saw the issue on emulator. Also, i am testing with both RTMP and RMPS, I hope RTMPS is not an issue. So, far, i am able to see the video but loosing the audio..trying to figure out why audio is not getting broadcasted. But the library seems to be super :)

I/Pedro: ----------------
    Name: c2.android.aac.decoder
    Type: audio/mp4a-latm
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Profile: 2, level: 0
    Profile: 5, level: 0
    Profile: 29, level: 0
    Profile: 23, level: 0
    Profile: 39, level: 0
    Profile: 20, level: 0
    Profile: 42, level: 0
    Bitrate range: 8000 - 510000
    Channels supported: 8
    Supported sample rate: 
    7350
    8000
    11025
    12000
    16000
    22050
    24000
    32000
    44100
    48000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.aac.decoder
    Type: audio/mp4a-latm
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Profile: 2, level: 0
    Profile: 5, level: 0
    Profile: 29, level: 0
    Profile: 23, level: 0
    Profile: 39, level: 0
    Profile: 20, level: 0
    Profile: 42, level: 0
    Bitrate range: 8000 - 510000
    Channels supported: 8
    Supported sample rate: 
    7350
    8000
    11025
    12000
    16000
    22050
    24000
    32000
    44100
    48000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.aac.encoder
    Type: audio/mp4a-latm
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 0
    Quality range: 0 - 0
    CBR supported: false
    VBR supported: true
    CQ supported: false
    ----- -----
    ----- Audio info -----
    Profile: 2, level: 0
    Profile: 5, level: 0
    Profile: 29, level: 0
    Profile: 23, level: 0
    Profile: 39, level: 0
    Bitrate range: 8000 - 510000
    Channels supported: 6
    Supported sample rate: 
    8000
    11025
    12000
    16000
    22050
    24000
    32000
    44100
    48000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.aac.encoder
    Type: audio/mp4a-latm
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 0
    Quality range: 0 - 0
    CBR supported: false
    VBR supported: true
    CQ supported: false
    ----- -----
    ----- Audio info -----
    Profile: 2, level: 0
    Profile: 5, level: 0
    Profile: 29, level: 0
    Profile: 23, level: 0
    Profile: 39, level: 0
    Bitrate range: 8000 - 510000
    Channels supported: 6
    Supported sample rate: 
    8000
    11025
    12000
    16000
    22050
    24000
    32000
    44100
    48000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.amrnb.decoder
    Type: audio/3gpp
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 4750 - 12200
    Channels supported: 1
    Supported sample rate: 
    8000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.amrnb.decoder
    Type: audio/3gpp
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 4750 - 12200
    Channels supported: 1
    Supported sample rate: 
    8000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.amrnb.encoder
    Type: audio/3gpp
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 0
    Quality range: 0 - 0
    CBR supported: true
    VBR supported: false
    CQ supported: false
    ----- -----
    ----- Audio info -----
    Bitrate range: 4750 - 12200
    Channels supported: 1
    Supported sample rate: 
    8000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.amrnb.encoder
    Type: audio/3gpp
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 0
    Quality range: 0 - 0
    CBR supported: true
    VBR supported: false
    CQ supported: false
    ----- -----
    ----- Audio info -----
    Bitrate range: 4750 - 12200
    Channels supported: 1
    Supported sample rate: 
    8000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.amrwb.decoder
    Type: audio/amr-wb
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 6600 - 23850
    Channels supported: 1
    Supported sample rate: 
    16000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.amrwb.decoder
    Type: audio/amr-wb
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 6600 - 23850
    Channels supported: 1
    Supported sample rate: 
    16000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.amrwb.encoder
    Type: audio/amr-wb
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 0
    Quality range: 0 - 0
    CBR supported: true
    VBR supported: false
    CQ supported: false
    ----- -----
    ----- Audio info -----
    Bitrate range: 6600 - 23850
    Channels supported: 1
    Supported sample rate: 
    16000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.amrwb.encoder
    Type: audio/amr-wb
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 0
    Quality range: 0 - 0
    CBR supported: true
    VBR supported: false
    CQ supported: false
    ----- -----
    ----- Audio info -----
    Bitrate range: 6600 - 23850
    Channels supported: 1
    Supported sample rate: 
    16000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.flac.decoder
    Type: audio/flac
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 1 - 21000000
    Channels supported: 8
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.flac.decoder
    Type: audio/flac
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 1 - 21000000
    Channels supported: 8
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.flac.encoder
    Type: audio/flac
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 8
    Quality range: 0 - 0
    CBR supported: false
    VBR supported: false
    CQ supported: true
    ----- -----
    ----- Audio info -----
    Bitrate range: 1 - 21000000
    Channels supported: 2
    ----- -----
    Max instances: 32
    ----------------
I/Pedro: ----------------
    Name: OMX.google.flac.encoder
    Type: audio/flac
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 8
    Quality range: 0 - 0
    CBR supported: false
    VBR supported: false
    CQ supported: true
    ----- -----
    ----- Audio info -----
    Bitrate range: 1 - 21000000
    Channels supported: 2
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.g711.alaw.decoder
    Type: audio/g711-alaw
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 64000 - 64000
    Channels supported: 6
    Supported sample rate: 
    8000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.g711.alaw.decoder
    Type: audio/g711-alaw
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 64000 - 64000
    Channels supported: 6
    Supported sample rate: 
    8000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.g711.mlaw.decoder
    Type: audio/g711-mlaw
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 64000 - 64000
    Channels supported: 6
    Supported sample rate: 
    8000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.g711.mlaw.decoder
    Type: audio/g711-mlaw
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 64000 - 64000
    Channels supported: 6
    Supported sample rate: 
    8000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.mp3.decoder
    Type: audio/mpeg
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 8000 - 320000
    Channels supported: 2
    Supported sample rate: 
    8000
    11025
    12000
    16000
    22050
    24000
    32000
    44100
    48000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.mp3.decoder
    Type: audio/mpeg
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 8000 - 320000
    Channels supported: 2
    Supported sample rate: 
    8000
    11025
    12000
    16000
    22050
    24000
    32000
    44100
    48000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.opus.decoder
    Type: audio/opus
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 6000 - 510000
    Channels supported: 8
    Supported sample rate: 
    8000
    12000
    16000
    24000
    48000
    ----- -----
    Max instances: 32
    ----------------
I/Pedro: ----------------
    Name: OMX.google.opus.decoder
    Type: audio/opus
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 6000 - 510000
    Channels supported: 8
    Supported sample rate: 
    8000
    12000
    16000
    24000
    48000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.opus.encoder
    Type: audio/opus
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 10
    Quality range: 0 - 0
    CBR supported: true
    VBR supported: true
    CQ supported: false
    ----- -----
    ----- Audio info -----
    Bitrate range: 6000 - 510000
    Channels supported: 2
    Supported sample rate: 
    8000
    12000
    16000
    24000
    48000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.raw.decoder
    Type: audio/raw
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 1 - 10000000
    Channels supported: 8
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.raw.decoder
    Type: audio/raw
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 1 - 10000000
    Channels supported: 8
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: c2.android.vorbis.decoder
    Type: audio/vorbis
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 32000 - 500000
    Channels supported: 8
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.vorbis.decoder
    Type: audio/vorbis
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 32000 - 500000
    Channels supported: 8
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.gsm.decoder
    Type: audio/gsm
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Audio info -----
    Bitrate range: 13000 - 13000
    Channels supported: 1
    Supported sample rate: 
    8000
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.android.goldfish.h264.decoder
    Type: video/avc
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    Profile: 65536, level: 65536
    Profile: 1, level: 65536
    Profile: 2, level: 65536
    Profile: 524288, level: 65536
    Profile: 8, level: 65536
    Bitrate range: 1 - 120000000
    Frame rate range: 1 - 480
    Width range: 96 - 3840
    Height range: 96 - 2160
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.android.goldfish.vp8.decoder
    Type: video/x-vnd.on2.vp8
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    Bitrate range: 1 - 100000000
    Frame rate range: 1 - 480
    Width range: 144 - 3840
    Height range: 144 - 2160
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.android.goldfish.vp9.decoder
    Type: video/x-vnd.on2.vp9
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    Profile: 1, level: 256
    Profile: 4, level: 256
    Profile: 4096, level: 256
    Profile: 16384, level: 256
    Bitrate range: 1 - 60000000
    Frame rate range: 1 - 480
    Width range: 144 - 3840
    Height range: 144 - 2160
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.h264.decoder
    Type: video/avc
    Max instances: 16
    ----- Decoder info -----
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    Profile: 65536, level: 65536
    Profile: 1, level: 65536
    Profile: 2, level: 65536
    Profile: 524288, level: 65536
    Profile: 8, level: 65536
    Bitrate range: 1 - 120000000
    Frame rate range: 1 - 480
    Width range: 2 - 2560
    Height range: 2 - 2560
    ----- -----
    Max instances: 16
    ----------------
    ----------------
    Name: OMX.google.h263.decoder
    Type: video/3gpp
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    Profile: 1, level: 4
    Profile: 1, level: 16
    Profile: 8, level: 4
    Profile: 8, level: 16
    Bitrate range: 1 - 384000
    Frame rate range: 1 - 30
    Width range: 4 - 352
    Height range: 4 - 288
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.hevc.decoder
    Type: video/hevc
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    Profile: 1, level: 65536
    Profile: 4, level: 65536
    Bitrate range: 1 - 10000000
    Frame rate range: 0 - 960
    Width range: 2 - 2048
    Height range: 2 - 2048
    ----- -----
    Max instances: 32
    ----------------
I/Pedro: ----------------
    Name: OMX.google.mpeg4.decoder
    Type: video/mp4v-es
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    Profile: 1, level: 16
    Bitrate range: 1 - 384000
    Frame rate range: 12 - 60
    Width range: 2 - 352
    Height range: 2 - 288
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.vp8.decoder
    Type: video/x-vnd.on2.vp8
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    Bitrate range: 1 - 100000000
    Frame rate range: 1 - 480
    Width range: 2 - 2560
    Height range: 2 - 2560
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.vp9.decoder
    Type: video/x-vnd.on2.vp9
    Max instances: 32
    ----- Decoder info -----
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    Profile: 1, level: 256
    Profile: 4, level: 256
    Profile: 4096, level: 256
    Profile: 16384, level: 256
    Bitrate range: 1 - 60000000
    Frame rate range: 1 - 480
    Width range: 2 - 2560
    Height range: 2 - 2560
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.h264.encoder
    Type: video/avc
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 0
    Quality range: 0 - 0
    CBR supported: false
    VBR supported: true
    CQ supported: false
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    21
    2130708361
    Profile: 65536, level: 4096
    Profile: 1, level: 4096
    Profile: 2, level: 4096
    Bitrate range: 1 - 20000000
    Frame rate range: 0 - 960
    Width range: 16 - 1920
    Height range: 16 - 1088
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.h263.encoder
    Type: video/3gpp
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 0
    Quality range: 0 - 0
    CBR supported: false
    VBR supported: true
    CQ supported: false
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    21
    2130708361
    Profile: 1, level: 16
    Bitrate range: 1 - 128000
    Frame rate range: 1 - 15
    Width range: 176 - 176
    Height range: 144 - 144
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.mpeg4.encoder
    Type: video/mp4v-es
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 0
    Quality range: 0 - 0
    CBR supported: false
    VBR supported: true
    CQ supported: false
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    21
    2130708361
    Profile: 4, level: 8
    Bitrate range: 1 - 64000
    Frame rate range: 12 - 60
    Width range: 16 - 176
    Height range: 16 - 144
    ----- -----
    Max instances: 32
    ----------------
    ----------------
    Name: OMX.google.vp8.encoder
    Type: video/x-vnd.on2.vp8
    Max instances: 32
    ----- Encoder info -----
    Complexity range: 0 - 0
    Quality range: 0 - 0
    CBR supported: true
    VBR supported: true
    CQ supported: false
    ----- -----
    ----- Video info -----
    Supported colors: 
    2135033992
    19
    21
    2130708361
    Profile: 1, level: 1
    Profile: 1, level: 2
    Profile: 1, level: 4
    Profile: 1, level: 8
    Bitrate range: 1 - 40000000
    Frame rate range: 0 - 960
    Width range: 2 - 2048
    Height range: 2 - 2048
    ----- -----
    Max instances: 32
    ----------------

**Comment by pedroSG94** [comment]
You should be able to stream because you have OMX.google.h264.encoder. This could be a bug.
- What do you mean with "i saw some success"? is it failing few times? can you reproduce this error 100% or only few times? (explain me the steps to reproduce)
- Can you tell me emulator specifications to reproduce the problem? model, android api version (specify if google play or google api room), ram, arquitecture, etc.
- Let me know if you are doing any modification in app example or you are using my app example without modifications (include modifications if used)
- Make sure that you are using app example in last commit (master branch)

**Reply by hemantkathuria** [reply]
What do you mean with "i saw some success"? is it failing few times? can you reproduce this error 100% or only few times? (explain me the steps to reproduce)

_In the previous test, i saw some audio loss when I used rtmps url instead of rtmp. But now, i have done the test again on another phone and it is broadcasting the audio. I will try again on the same phone tomorrow as i dont have it with me._

Can you tell me emulator specifications to reproduce the problem? model, android api version (specify if google play or google api room), ram, arquitecture, etc.

_Please find below as attachment_

Let me know if you are doing any modification in app example or you are using my app example without modifications (include modifications if used)
Make sure that you are using app example in last commit (master branch)

_Currently no changes in the code, i was trying it on multiple devices to test._ 

[Emulator Details.txt](https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/files/7520650/Emulator.Details.txt)
)

**Comment by AlexUrrutia** [comment]
Hello Pedro, I am experiencing the same issue, using an emulator and the default app from your repository, I also have reports from users with different type of devices having the problem, not sure where to go from there

`2021-11-12 11:40:19.424 5640-5640/com.pedro.rtpstreamer E/VideoEncoder: Valid encoder not found`

**Emulator details:**
<img width="1195" alt="Screen Shot 2021-11-12 at 11 43 04 AM" src="https://user-images.githubusercontent.com/7042361/141502954-3c048654-491a-41ff-bcba-68d1b8b8ede4.png">

**Comment by pedroSG94** [comment]
Fixed:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/commit/5d07e657ea725467e669c573a59cf1b08196bac6
You can update to version 2.1.4

**Comment by AlexUrrutia** [comment]
Yes, it is working now, thank you!

**Comment by pedroSG94** [comment, marked as answer]
@hemantkathuria please, mark as answered to avoid confusion
